fix(core): stricter event emitter type

amakhrov opened the pull request:
This completes the work from this original PR: https://github.com/angular/angular/pull/35300 (no longer maintained).

However, I discarded the original idea to default `new EventEmitter()` to `EventEmitter<void>`, instead sticking to the existing type of `EventEmitter<any>`. That would be yet another breaking change with way too many cascading changes to make in tests.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Issue Number: Fixes #35119

This error is not caught and can cause a runtime exception, since a subscriber only expects values of type `number` to be emitted.
```
const em = new EventEmitter<number>;
em.next(); // <- No value provided will feed `undefined` where `number` is expected
```

## What is the new behavior?

This error is caught by TypeScript
```
const em = new EventEmitter<number>;
em.next(); // <- Expected 1 argument to `next()` but 0 arguments provided
```

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

It's no longer possible to have an event emitter type which sometimes accepts arguments and sometimes not. It's either one but not both:
```
const em = new EventEmitter<void>();
em.next(); // OK
em.next(something); // Not OK

const em2 = new EventEmitter<any>();
em2.next(); // Not OK
em2.next(undefined); // OK
em2.next(something); // OK
```
This could be overcome if `emit` is defined as a typed property rather than a method. [Example in TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogtgS2AHgCoD4oF4oCgoFSpQQAewEAdgCYDOUAbgPYLX6EcD8UAFAJTZMzVuw4EAXFACuNCADMElCNRLkqdIqLFdeDAIYAbKRE6TUArEJZtt2yT31GIZi1ZG5QkWAyrB4SCgAnNEwcAG8tCERgSX8UDABuXABfXFwDCGASOABCWJ9KP2ig5D1KEHQkqIA6KKR+IA). However, I'm reluctant with such an approach - announcing it as "property" - might be confusing to some users.

Any feedback on this subject is welcome!

## Broken consumer in angular/components

The breaking change described above currently breaks an example in material-experimental: https://github.com/angular/components/blob/master/src/components-examples/material-experimental/popover-edit/popover-edit-mat-table/popover-edit-mat-table-example.html#L100

What is a proper way to proceed with it? Should a separate PR be filed for angular/components ? 

## Other information
Original PR: https://github.com/angular/angular/pull/35300
Related PR in rxjs: https://github.com/ReactiveX/rxjs/pull/5307

amakhrov commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -2025,7 +2025,7 @@ describe('di', () => {
       class MyDir {
         @Input() binding!: string;
-        @Output() output = new EventEmitter();
+        @Output() output = new EventEmitter<any>();
```
this and some other tests were updated (to explicitly specify the type) as a part of the original PR, which introduced the default of `void` instead of `any`.

this is not longer required, since I restored the `any` default. However, I believe keeping it explicit won't hurt any way, and can be beneficial in future if we finally decide to change the default or that the generic parameter must always be specified

amakhrov replied:
reverted, as per comment from @AndrewKushnir

AndrewKushnir commented:
Hi @amakhrov, thanks for creating this PR. It looks like there is a problem with CLA due to co-authoring commits (which you also mentioned in your comment https://github.com/angular/angular/pull/35300#issuecomment-695861910). Since you re-worked the implementation, may be it makes sense to create a new commit and have a reference to a parent PR in commit description?

Also quick comment on the actual changes: I think we should remove generic where it's not strictly needed - that'd help keep the changes as minimal as possible and will also highlight the places where the generic is needed (so these cases can be checks additionally during the code review).

Thank you.

amakhrov committed:
fix(core): stricter event emitter type

Given `EventEmitter<T>`, its `.emit(value)` method accepts argument exactly of type T.
No longer allow to pass `undefined` or omit the argument.

BREAKING CHANGE:
- for `EventEmitter<any>`, its `.emit()` call (no argument) no longer works.
Use `.emit(undefined)` instead, or declare it as `EventEmitter<void>`;

Fixes #35119

AndrewKushnir commented:
@amakhrov, thanks for applying additional changes.

As you mentioned in PR description, this is a breaking change and I also confirmed it by running tests in Google's codebase. It looks like this change would also require a migration and we may not be able to look into that before v11 release. I'm adding this PR to the v12 candidates milestone and we'll get back to that once we start working on v12 items.

Thank you.

hmil commented:
One note about the default setting that you chose. If you chose to make `any` the default rather than `void`, then the argument-less form of `emit()` will not work for EventEmitters created without an explicit type parameter. This can cause more breaking changes than you'd think. A lot of the tests use `emit()` with EventEmitters created without parameters.

Whether `void` is more strongly breaking than `any` depends on which usage you consider to be the most prevalent usage.

I think a lot of people use explicit type parameters when they intend to use the emitter with a given type, especially for non-primitive types. However, I don't think anyone bothered to declare `EventEmitter<void>` the emitters which they intended to use only in argument-less form.
Therefore, it is almost guaranteed that people will have to refactor their argument-less emitters into `EventEmitter<void>`.

Another argument to consider is that of self documentation: Assuming a beginner level in TypeScript, if I copy an example which has `EventEmitter<string>` and uses `emit('some string')`, then it appears consistent that if I remove `<string>` from the declaration I can drop the argument at usage. This is the behavior with `void` as default, but not that with `any` as default.
With `any` as default, even advanced TypeScript users will fail to notice that they could specify `<void>` to allow the parameter-less usage form `emit()`.

As demonstrated in the tables below, the mental model with `any` as default is less intuitive than that with `void` as default.

Usage with `void` as default exhibits a nice parallelism which makes the mental model straightforward:

Declaration | Expected usage
---|---
EventEmitter&lt;string> | emit('some string')
EventEmitter&lt;number> | emit(42)
EventEmitter | emit()
EventEmitter&lt;any> | emit(anything)

Usage with `any` as default breaks parallelism for argument arity:

Declaration | Expected usage
---|---
EventEmitter&lt;string> | emit('some string')
EventEmitter&lt;number> | emit(42)
EventEmitter | emit(anything)
EventEmitter&lt;void> | emit()

---

While one may think that `any` is a more permissive and open default, it is not entirely true when you consider the argument-less version of `emit`. `any` is just as likely to break existing code as `void` is. Besides, using `any` as a default creates a counter-intuitive mental model.

At the end of the day it's your call since you make the effort to implement this. But I hope that at least you make this choice with a complete understanding of what's at stake. It is unlikely that the default will ever be changed, so better make sure to get it right from the start.

amakhrov commented:
@AndrewKushnir what do you think about more sophisticated typing that would allow to omit argument when `T` allowed `undefined` (e.g. `any`) - as in the typescript playground linked in the PR description?

While it does have an undesired effect of the `emit` being reported as `property` rather than `method`, it allows to eliminate a breaking change in this PR. I personally prefer a stricter and more straightforward approach as introduced in the PR. However, if getting rid of the breaking change unblocks this PR earlier, I'll totally support it.

hmil commented:
You can also try [this simpler alternative](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsGBbAlsYEBOAeAFQD4oBeKAbwCgpaoINgAKYALigFVEATCAMwIB7AGqD03QkQCU7OGO4BuKgF8qVUJE49+Q0eMlkoAV2190iCN3oAPbDwDOUAlAD8UOeKjsCSqgBsIYHpUAEJ2eCQURmx8AENEECIlBgA6BkwmKSA) which gives a similar result but keeps the method declared as such.

But I also prefer the more straightforward approach if we can afford a breaking change.

amakhrov commented:
> You can also try this simpler alternative 

It doesn't allow to pass arbitrary value for `EventEmitter<any>` this way.

AndrewKushnir commented:
Hi @amakhrov,

> @AndrewKushnir what do you think about more sophisticated typing that would allow to omit argument when T allowed undefined (e.g. any) - as in the typescript playground linked in the PR description?

I think the current set of changes is reasonable. I noticed that you've updated the implementation to use `void` instead of `any`, which makes sense to me based on https://github.com/angular/angular/pull/38916#issuecomment-696933337 and given that it would also make `EventEmitter` types closer to the parent `Subject` class types. That would still be a breaking change and given the impact we would need to consider creating a schematic to help migrate affected `EventEmitter` usages in apps code. We'll get back to this when we start working on the v12 preparations. Thank you.

kapunahelewong reviewed (changes requested):
Thanks for these updates, @amakhrov! I have to defer to the engineers to approve the code changes, but if you change `examples/getting-started/src/app/product-alerts/product-alerts.component.ts`, we need to keep the documentation updated as well and explain the use of `void` in the `EventEmitter`. The description of this code snippet is in the Getting Started here: https://angular.io/start#pass-data-to-a-parent-component.

Immediately after the `input-output` example snippet (line232) you could say something like:


> You can use the `void` type to specify that the `EventEmitter()` does not have a type.
> Using the `void` type  is a good practice because...
> 

Because it allows for returning no data, which avoids errors or minimizes bugs...? (I'm not sure exactly!) Let me know if you need any help with wording!

amakhrov commented:
@kapunahelewong I updated the readme to clarify why `void` is used.
I didn't mention that it's a good practice or anything. With the change introduced by the PR, it's mandatory to specify `void` when you use your event emitter without any values passed to .emit() (as in the example).

Perhaps more details could be added to API docs for EventEmitter class, though. Any thoughts?

kapunahelewong commented on aio/content/start/index.md:
```diff
@@ -231,6 +231,8 @@ The `ProductAlertsComponent` needs to emit an event when the user clicks **Notif
 
   <code-example path="getting-started/src/app/product-alerts/product-alerts.component.ts" header="src/app/product-alerts/product-alerts.component.ts" region="input-output"></code-example>
 
+  Note how the `void` type is used to specify that `notify.emit()` (below) won't accept any argument.
```
```suggestion
You use the `void` type to specify that `notify.emit()` doesn't accept an argument.
```

Just switches to the active from the passive voice.

kapunahelewong reviewed (changes requested):
Yes, this works for me, @amakhrov. I've added just a small comment to bring it in line with the style guide. Thank you so much!

kapunahelewong reviewed (dismissed):
LGTM! We'll need engineering for the code itself, but from a docs perspective it looks good! Thanks, @amakhrov!

Reviewed-for: global-docs-approvers

petebacondarwin commented on packages/examples/upgrade/static/ts/full/module.ts:
```diff
@@ -41,7 +41,7 @@ export class TextFormatter {
   @Input() heroes!: Hero[];
-  @Output() addHero = new EventEmitter();
+  @Output() addHero = new EventEmitter<void>();
   @Output() removeHero = new EventEmitter();
```
While not necessary for the typings to pass compilation, I think this should be changed to:

```suggestion
  @Output() removeHero = new EventEmitter<Hero>();
```

to set a good example, and to provide better typing in the template.

mary-poppins commented:
You can preview 2fe981b at https://pr38916-2fe981b.ngbuilds.io/.
You can preview fb63325 at https://pr38916-fb63325.ngbuilds.io/.
You can preview 25d77f8 at https://pr38916-25d77f8.ngbuilds.io/.
You can preview c81c89c at https://pr38916-c81c89c.ngbuilds.io/.
You can preview e5c4f8c at https://pr38916-e5c4f8c.ngbuilds.io/.
You can preview a9f3668 at https://pr38916-a9f3668.ngbuilds.io/.

petebacondarwin reviewed (changes requested):
I think we have to be careful with the API docs here...

If I understand correctly, the `EventEmitter` interface must always have a type parameter, while the `EventEmitter()` class constructor will result in either `EventEmitter<T>`, if a type param is provided, or `EventEmitter<any>`, if no type param is provided. AFAICT this has not changed from before this PR.

But now you must provide a type param if `emit()` will be called with no argument.

This is not at all clear from the API docs for `EventEmitter` - https://pr38916-a9f3668.ngbuilds.io/api/core/EventEmitter.

I think the `emit()` method should mention this; but also the overall class extended description should mention it too; and the inline example in that current description should probably be updated to have better typings. Currently the type of the `open` and `close` outputs is typed to `any`, while in fact they are only called with `null`, which in itself is redundant.

Perhaps this example should be changed to `void` for both outputs, which would also provide an opportunity to explain about how `void` is different from `any`?

Finally, I don't think we should land this PR until it has an associated migration included. There will be many breaking changes where the event emitter has no type param but where `next()` is called with no argument.  I am not sure how easy it would be to write this migration, since we cannot simply swap all `new EventEmitter()` expressions with `new EventEmitter<void>()`. We would need to look for places where the `emit()` method is called on each output, and work out from that what type to specify (if any [sic]).

amakhrov commented:
@petebacondarwin Great points!
- Agree api docs for `EventEmitter` should be improved to mention the type parameter (this could have been done before) and explain the `void` situation (new with this PR)
- Also agree migration would be ideal here. I have not idea on how to implement it though - any clues, starting point to look at?

Also I'm currently blocked here by https://github.com/angular/angular/issues/37329

petebacondarwin commented:
You can see all our previous migrations here: https://github.com/angular/angular/tree/master/packages/core/schematics/migrations

I am not sure which one would provide the best example as a starting point. I think the migration approach would be to use the TypeChecker to find all `EventEmitter.next` symbols. Then if there is one that calls this without any arguments, to find the initialization place for the callee and check that it has a `<void>` type param.

petebacondarwin commented:
@amakhrov - I believe you should be unblocked on #37329 now. Can you take a look at updating this PR?

amakhrov commented:
@petebacondarwin thanks - no more issues with Bazel for me!

I created a migration for a couple of most typical use cases. That is - event emitter defined as a property of class with an initializer, and used in the same class or in its template. I don't know how to properly utilize type checker for a parsed template - hence my approach here is naive and probably not the most robust one. For example, I can imagine it giving a false positive (i.e a migration is incorrectly applied) if there is smth like this in the component's template:
```
<ng-template let-emitter>
  <button (click)="emitter.emit()></button>
</ng-template>
```
Another notable miss (false negative this time) is using `EventEmitter` as a variable rather than a class property. E.g. this will not be migrated:
```
const myEmitter = new EventEmitter();
myEmitter.emit();
```

mary-poppins commented:
You can preview 6ce870e at https://pr38916-6ce870e.ngbuilds.io/.

petebacondarwin commented on packages/core/schematics/test/void_event_emitter_migration_spec.ts:
```diff
@@ -0,0 +1,251 @@
+        change = new EventEmitter<number>();
+
+        onClick() {
+          this.change.emit();
```
In this case this will result in a compilation error that the developer would have to resolve manually?

petebacondarwin replied:
Would it be worth flagging these as warnings during the migration (if that is possible)?

amakhrov replied:
Yep, it would be a compilation error.
There are different possible scenarios and ways to fix it, though - so I decided not to migrate automatically.

We can probably flag all cases of incorrect usage (void call for an emitter declared / initialized withg a non-void type parameter).
We can also flag "mixed" cases (as in the unit test below).

What is the proper approach for generating warnings? Just `console.log`? Or is there any standard way to do that in migrations?

amakhrov replied:
On the other hand - the warnings will generally repeat the corresponding TS compiler errors. Do you think it's beneficial to have them anyway?

petebacondarwin replied:
I feel it is better to flag them during migration. Otherwise it will look like the update worked and then the compilation will fail, which is a confusing DX.

I am not sure how best to surface logging messages.  Perhaps @alan-agius4 could say?

amakhrov replied:
I found some examples in the older migrations (e.g. https://github.com/angular/angular/blob/master/packages/core/schematics/migrations/missing-injectable/index.ts#L36-L38) - will use the same approach

petebacondarwin replied:
Perfect!

petebacondarwin commented on packages/core/schematics/test/void_event_emitter_migration_spec.ts:
```diff
@@ -0,0 +1,251 @@
+        change = new EventEmitter();
+
+        onClick() {
+          this.change.next();
```
How about a test with a mix of `next()` and `emit()` calls. What would happen if all of one are without args but the other does have args, for instance?

amakhrov replied:
makes sense, I'll update the tests to reflect that.

the idea is that the both methods are treated identically under the same bucket. it shouldn't matter whether a developer uses .next, or .emit, or the both at the same time.

petebacondarwin commented on packages/core/src/event_emitter.ts:
```diff
@@ -80,7 +80,7 @@ export interface EventEmitter<T> extends Subject<T> {
    * @param value The value to emit.
    */
-  emit(value?: T): void;
+  emit(value: T): void;
```
Could we update the docs for this to make it clear what it means that `value:T` is not optional; and to give the example of setting `T` to `void` in the case where no arg is expected?

amakhrov replied:
Should we change the `Zippy` example class in the tsdoc section to use `void` event emitters? The current example looks quite dated, I don't believe it represents the best practice.

petebacondarwin replied:
Do you think that this update would need to happen in this PR? Or if it is just cleaning up the docs to make it less dated then perhaps it could happen in a later PR?

amakhrov commented on packages/core/schematics/migrations.json:
```diff
@@ -94,6 +94,11 @@
       "description": "`XhrFactory` has been moved from `@angular/common/http` to `@angular/common`.",
       "factory": "./migrations/xhr-factory/index"
+    },
+    "migration-v13-void-event-emitter": {
```
I believe I'm late to the v12 release party - so targeting this for v13

hmil commented:
@amakhrov You may want to [try this](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcCyBDRAjYB5MGASwgDsMAbABQygwFtgZgoAeAFQD44BeOAV1IATYADMipYELihmwgM5x2cAPxwA2uxUBdOAC4N7bQG4AUKaQo4AUQBuwUjGv0iMZmy684Ab1Nx-Mi4wABQAdOHqMNoGmDj4hCTk1LQMTCwcnACUBrYQREJmAXAA9MWBrsEwBuzZcLn5ZgC+5paoAKrCYuwQAGp5QhlegiLiktKyDkKKymr10tVmphRMgQCEBnYOTkHurBikiJxmwPShJxUA5BgXmaYnZ0HBt-fnIcNiElK3SysnAEzrGz2RzOVy7eQwKASADmRzu9D+D0u12eCKRIVRiNewXeoy+5mW8BOAGZAZsQTt0rjPkI4ST0cErjd4cSGajWdjqWNMkA) for the type signature of emit (credits to @spenceryue ).

Similar to the suggestion I made a few month ago, but without the downside [which you pointed out](https://github.com/angular/angular/pull/38916#issuecomment-696979312). It should behave like so:

- with `<any>`: allows `emit()` and `emit(anything)`
- with `<void>` or `<undefined>`: allows `emit()` and `emit(undefined)`
- with `<T>` where T is a _regular_ type: allows only `emit(T)`

amakhrov commented:
@hmil thanks, that looks great. It would allow to avoid a breaking change and won't require a migration.

The only drawback I still see is that the method signature is used in the autogenerated api docs and for context help from language service. Typehints become less straightforward:
<img width="437" alt="TypeScript__TS_Playground_-_An_online_editor_for_exploring_TypeScript_and_JavaScript" src="https://user-images.githubusercontent.com/1004015/118524741-538d2800-b6f3-11eb-9548-38c6f3220398.png">

@petebacondarwin do you think it could be good approach?

petebacondarwin commented:
@amakhrov - I'll discuss with @AndrewKushnir and get back to you.

spenceryue commented:
> The only drawback I still see is that the method signature is used in the autogenerated api docs and for context help from language service. Typehints become less straightforward

@amakhrov In TS 4.0, [labelled tuple elements](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements) were added to address the disparity between functions declared with an ordinary parameter list and functions declared with a [rest parameter with tuple type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#rest-parameters-with-tuple-types).

I've updated @hmil's [playground](https://www.typescriptlang.org/play?ssl=16&ssc=4&pln=16&pc=6#code/C4TwDgpgBAsghiARhA8mYBLA9gOzgGwAUF8s4ATAHgBUA+KAXigFcdyIAzDHCcqCAB7AIbAM5RqUAPxQA2lnTY8REmXJSAXBIC6ULbLCqKW6toDcAKAvsAxvjgAnaHbijxAZWaIAVhBvAaegBvCygwqAB6CKgAOjiLAF8rFzcoAFEANxFgNIBbDGBhBxpGKDgcEHpBYTEoTx8-ALooEPCoMC98DBt+fOAACjiYgyNybS14JFRFXAJiEFIKQIBKLQysDD5WtvCo2Pi2pKSrW3snKBtcUWB+AAYtTOy8gqLKcsrLCFuYiD7+gHI4P9lpY9oBeDcAYjtQfq5CDAAAWWHIy3SWRwOT6r3etB+fwUmFmKgWak0ZQqqyg602Fi+uIK-RBNO+v3prHYXB4yMs1j8Z2cVxuEAAjA80RiXhBitcHNwAOa0T5CukDQHAxXKhmg6IAPV1evSAkg-l4UCFZQcsuYsPRogANFBEMwbrKsDdvv0AEwAVi9ABZljSlSyBmzONxeIy9js9THdVAAIIWq3ZKBYDhQUCQKD-UMc3j-KAYcQ4V1lNwYWV4RD4aDALDtRxwWFFVPpzPQf7SuX-GKegDMvq9AZ5LnOlxw134HtFT0xksoufD5AVNI9GtVkei0bjict1puaYz4A7ACI4CeC0WoCWbq5RBWqzWM-XDA4m3DJa2j1mc2ww5ye37QcAwgNdg01VcNUXTlGSsKJ4IQxCkL2ABxERJTgYQ+BicgYmAcRkMI5CLD2IYSOiU5HGcexUkedFnkKedJCYbF+CEERyA8LxfH8QIWnInZwKGEZiQocZYAQZA0AJZR5kWKg6ApKlyC1KAklIuIgA) to use tuple element labels:
```typescript
type MaybeOptionalPayload<T> = undefined extends T ? [optionalPayload?: T] : [payload: T];

declare class Subject<T> {
    // ...
}

class EventEmitter<T = any> extends Subject<T> {
    public emit(...[payload]: MaybeOptionalPayload<T>): void {
        // ...
    }
}


declare const e0: EventEmitter<any>;
e0.emit('a');
// 👆 (method) EventEmitter<any>.emit(optionalPayload?: any): void
e0.emit();
e0.emit(undefined);

declare const e1: EventEmitter<string>;
e1.emit('a');
e1.emit();
// ^^^^^^ Expected 1 arguments, but got 0.(2554)
e1.emit(undefined);
//      ^^^^^^^^^ Argument of type 'undefined' is not assignable to parameter of type 'string'.(2345)

declare const e2: EventEmitter<undefined>;
e2.emit('a');
//      ^^^ Argument of type '"a"' is not assignable to parameter of type 'undefined'.(2345)
e2.emit();
e2.emit(undefined);


////////////////////// Generated .d.ts //////////////////////
// ...
// declare class EventEmitter<T = any> extends Subject<T> {
//     emit(...[payload]: MaybeOptionalPayload<T>): void;
// }
// ...
```

amakhrov commented:
@petebacondarwin any further thoughts on the approach here?

petebacondarwin commented:
Hi @amakhrov - thanks for the push. @AndrewKushnir and I discussed this today. We think it would be interesting to investigate the last idea with the labelled tuple elements to see how ergonomic that would be, both in terms of IDE autocompletion and angular.io documentation. 

If you would like to try this out so we can see it concretely, that would be great. But to be completely open, we may well decide that this would not be worth the API churn.

amakhrov commented:
Good point on IDE autocompletion @petebacondarwin .
There are existing open issues for WebStorm: https://youtrack.jetbrains.com/issue/WEB-49488

WebStorm does not properly unfold the tuple elements when showing autocompletion hints, falling back to `MaybeOptionPayload`:

![Screen Shot 2021-06-10 at 2 08 43 PM](https://user-images.githubusercontent.com/1004015/121598006-2e3fc100-c9f6-11eb-8787-289cbab66ffb.png)

![Screen Shot 2021-06-10 at 2 08 53 PM](https://user-images.githubusercontent.com/1004015/121598004-2e3fc100-c9f6-11eb-9306-5ca7cdb00c03.png)

But for already written code it shows correct hints:
![Screen Shot 2021-06-10 at 2 16 25 PM](https://user-images.githubusercontent.com/1004015/121598274-84146900-c9f6-11eb-963b-efd1bd00048a.png)

I really expect IDE support to improve over time :)

petebacondarwin commented:
@amakhrov - did you have time to look into the labelled tuple approach?

amakhrov commented:
@petebacondarwin 
Well, as mentioned above, WebStorm support still has some space for improvement with regards to that.
And in the public api docs it looks like this:
```
export interface EventEmitter<T> extends Subject<T> {
  emit(...[maybeOptionalPayload]: MaybeOptionalPayload<T>): void;
}
```
I'm not even sure how to properly annotate it in TSDoc as `@param`

amakhrov commented:
Honestly, I think we need to follow the approach RxJS has taken. After all, EventEmitter extends Subject. If we diverge (even with a good intention of minimizing migration friction) - it will add confusion in future.

petebacondarwin commented:
> Honestly, I think we need to follow the approach RxJS has taken. After all, EventEmitter extends Subject. If we diverge (even with a good intention of minimizing migration friction) - it will add confusion in future.

Can you clarify what the RxJS approach is here?

alxhub commented:
@amakhrov are you able to rebase this PR and mark it as ready for review? This is something we'd like to try and land in Angular 14.

amakhrov commented:
@alxhub do you meant the approach currently implemented in the PR (with somewhat partial migrations added)? Or the approach proposed by @spenceryue https://github.com/angular/angular/pull/38916#issuecomment-843211513 ?

mary-poppins commented:
You can preview 5b90024 at https://pr38916-5b90024.ngbuilds.io/.

mary-poppins commented:
You can preview a054908 at https://pr38916-a054908.ngbuilds.io/.

mary-poppins commented:
You can preview 80a4a87 at https://pr38916-80a4a87.ngbuilds.io/.

amakhrov commented:
Note that this is still not resolved: https://github.com/angular/angular/pull/38916#discussion_r626092268 . Requires some extra work - if we're sure we want to move on with the PR

dylhunn commented:
Hi all! I notice that I've been added as a reviewer on this PR.

@amakhrov Is this something we want to merge soon, and if so, are we going to pursue the original breaking approach? If so, we will need to add a migration. New migrations should be placed in [this directory](https://github.com/angular/angular/tree/master/packages/core/schematics/migrations). They're actually quite easy to write, so don't be put off by the need to add one.

amakhrov commented:
@dylhunn thanks!

Actually the migration is already there.

What's missing, as per the thread with Pete above, is displaying warnings in cases where the code wasn't migrated because of ambiguity (basically, cases where the existing code violates strict null checks).
I tried to add that quickly - but turned out it would require some significant changes in the migration script, so I put it on a back burner.

Do you think it still makes sense to add it?

dylhunn commented:
We just discussed this at triage meeting, and we do want to move ahead. @atscott noted that a similar migration would be useful for rxjs `Subject`.

amakhrov commented:
@dylhunn just to clarify - would you like me add warning messages to the migration script? Or do you plan to move on with it as is?

Anything else needed in this PR?

dylhunn commented:
@amakhrov Let's add the warning messages, and I (and/or future weekly caretakers) can shepherd the migration through. Would you also mind rebasing? Thanks!

mary-poppins commented:
You can preview 485db63 at https://pr38916-485db63.ngbuilds.io/.

mary-poppins commented:
You can preview 2d43a08 at https://pr38916-2d43a08.ngbuilds.io/.

amakhrov commented:
@dylhunn ready.

The failing "components-repo-unit-tests" stage reflects a problem in the material repo (strict null check error, flagged because the map is [initialized](https://github.com/angular/components/blob/7dbe9a42829a2e71cca3363e9a97ec0ea45be276/src/google-maps/google-map/google-map.ts#L308-L312) inside a callback).

dylhunn commented:
@amakhrov Thanks! This will require us to run the migration inside of google before we can merge it. I blocked off some time for next week to work on that with Jess, and we'll get back to you with an update.

dylhunn commented:
Good morning @amakhrov! @jessicajaniuk  and I just met to look at this PR, and figure out the next steps for the migration. Unfortunately, we noticed that there is no TSLint version of the migration, for use in google3. Would you be able to add one for us?

Here's an example PR that includes the TSLint migration: [Typed Forms Migration](github.com/angular/angular/pull/44449)

It's not too difficult to add -- basically, you'll just need another file that calls into your existing `index.ts` and `util.ts` helpers, and returns the failures as TSLint violations. You'll want to also return the warnings through TSLint, although I'm not sure what the exact right format is. You should be able to run the tests locally as well -- even though it's in a folder called `google3`, all the test running code is in the repo.

Once that is done we'll be able to run the schematic on google3, and report back with info about the extent of non-migratable code.

Thanks for your contribution, and effort towards fixing this longstanding bug!

dylhunn commented:
Some additional context: [link1](https://github.com/ReactiveX/rxjs/pull/5389), [link2](https://source.corp.google.com/piper///depot/google3/third_party/javascript/rxjs/src/internal/Subject.ts;l=81)

We are in a pretty weird situation, where we support both RxJS6 and the latest version -- Subject is fixed at head, but not in 6. Google actually uses a local mod of RxJS to add back the optionality, to avoid breaking targets.

If this change proves excessively breaky, we should consider doing the same with a local mod for Google only.

amakhrov commented:
@dylhunn I'll take a stab on the tslint migration

dylhunn commented:
Thanks @amakhrov, I'll be out until Monday, but happy to have a look or help once I'm back.

amakhrov commented:
@dylhunn The migration flags incorrect calls of `.emit()` method (as warnings). Such calls can occur in a template file (.html).  This file isn't a part of a TS AST (it's resolved from the decorator metadata), and doesn't have a corresponding ts.SourceFile node. It means I cannot create a tslint RuleFailure object for that.

Any hints on how to approach it? Should I flag the emitter declaration in the class instead of flagging the incorrect usage of .emit()?

I'm talking about the case covered in this unit test: https://github.com/angular/angular/pull/38916/files#diff-e8d7f17e13abd7a83565b5ebb9909ed13a5ac7370a60bc1ceac03168d223c3afR298-R314

dylhunn commented:
> Should I flag the emitter declaration in the class instead of flagging the incorrect usage of .emit()?

This might be a reasonable workaround. It's definitely not quite as nice though, since there could be many usages of that declaration. The TSLint version only affects us, so as long as it's reasonably quick to track down the culprit, it should be alright.

dylhunn commented:
@MichaelJamesParsons We found this PR while triaging. Is this something you would need for the RxJS migration?

JeanMeche commented:
Hi, I'll close this PR as we have now a better alternative with the view queries which fix this typing issue.
